src/net/tls.cc

```diff
@@ -1980,6 +1981,9 @@ std::ostream& tls::operator<<(std::ostream& os, const subject_alt_name& a) {
     return os << a.type << "=" << a.value;
 }
 
+void* tls::get_impl(const tls::certificate_credentials &creds) {
+    return creds._impl->_creds;
```

Comment:
Why return void*? What is the caller supposed to do with it?

Reply:
operator is too magical, I prefer the explicit `as_gnutls_certificate_credentials_t`.

about abstracted accessors, I agree in principle, but if this a shallow-but-wide interface (large number of simple getters) it may be difficult to maintain.

Reply:
(is it a shallow-but-wide interface?)

Reply:
> is it a shallow-but-wide interface?

Not really. The information we want is:
- serial number and expiration time of each loaded cert
- serial number and expiration time of each CA in the trust list

I have a working branch that adds the following interface to `net::tls::certificate_credentials`:

```c++
std::optional<std::vector<cert_info>> get_cert_info() const noexcept;
std::optional<std::vector<cert_info>> get_trust_list_info() const noexcept;
```

Reply:
If that looks agreeable, I'll get my branch cleaned up resubmitted here.

For reference, `cert_info` looks like this:
```
struct cert_info {
    sstring serial;
    time_t expiry;
};
```

Reply:
Sure it makes sense. Presumably, the same interface would work for openssl?

Reply:
I assume so. The gnutls bits are handled inside `::impl`, but I don't know the corresponding openssl incantations offhand.